src/LanguageServer/Protocol/Extensions/SourceGeneratedDocumentUri.cs

```diff
@@ -0,0 +1,112 @@
+            var hintPathPortion = string.Join("/", hintPathParts.Select(Uri.EscapeDataString));
+
+            var uri = Scheme + "://" +
+                identity.DocumentId.ProjectId.Id.ToString(GuidFormat) + "/" +
+                hintPathPortion +
+                "?documentId=" + identity.DocumentId.Id.ToString(GuidFormat) +
+                "&hintName=" + Uri.EscapeDataString(identity.HintName) +
+                "&assemblyName=" + Uri.EscapeDataString(identity.Generator.AssemblyName) +
+                "&assemblyVersion=" + Uri.EscapeDataString(identity.Generator.AssemblyVersion.ToString()) +
+                "&typeName=" + Uri.EscapeDataString(identity.Generator.TypeName);
+
+            // If we have a path (which is technically optional) also append it
+            if (identity.Generator.AssemblyPath != null)
+                uri += "&assemblyPath=" + Uri.EscapeDataString(identity.Generator.AssemblyPath);
+
+            return ProtocolConversions.CreateAbsoluteUri(uri);
+        }
+
+        public static SourceGeneratedDocumentIdentity? DeserializeIdentity(Solution solution, Uri documentUri)
+        {
+            // This is a generated document, so the "host" portion is just the GUID of the project ID; we'll parse that into an ID and then
+            // look up the project in the Solution. This relies on the fact that technically the only part of the ID that matters for equality
+            // is the GUID; looking up the project again means we can then recover the ProjectId with the debug name, so anybody looking at a crash
+            // dump sees a "normal" ID. It also means if the project is gone we can trivially say there are no usable IDs anymore.
+            var projectIdGuidOnly = ProjectId.CreateFromSerialized(Guid.ParseExact(documentUri.Host, GuidFormat));
+            var projectId = solution.GetProject(projectIdGuidOnly)?.Id;
+
+            if (projectId == null)
+                return null;
+
+            Guid? documentIdGuid = null;
+            string? hintName = null;
+            string? assemblyName = null;
+            string? assemblyPath = null; // this one is actually OK if it's null, since it's optional
+            Version? assemblyVersion = null;
+            string? typeName = null;
+
+            // Parse the query string apart and grab everything from it
+            foreach (var part in documentUri.Query.TrimStart('?').Split('&'))
+            {
+                var equals = part.IndexOf('=');
+                Contract.ThrowIfTrue(equals <= 0);
+#if NET
+                var name = part.AsSpan()[0..equals];
+#else
+                var name = part.Substring(0, equals);
+#endif
+                var value = Uri.UnescapeDataString(part.Substring(equals + 1));
+
+                if (name.Equals("documentId", StringComparison.Ordinal))
+                    documentIdGuid = Guid.ParseExact(value, GuidFormat);
+                else if (name.Equals("hintName", StringComparison.Ordinal))
```

Comment:
these should be constnats, if we're using it to both create and break up the uri.